src-ui.v3/src/Main.hs

```diff
@@ -96,337 +98,381 @@ utc2unix x = ceiling (realToFrac (utcTimeToPOSIXSeconds x) :: Double)
 
 bodyElement4 :: forall t m . (SupportsServantReflex t m, MonadFix m, MonadIO m, MonadHold t m, PostBuild t m, DomBuilder t m, Adjustable t m, DomBuilderSpace m ~ GhcjsDomSpace) => m ()
 bodyElement4 = do
-    dynLoc <- browserHistoryWith getLocationUri
-    let dynFrag = decodeFrag . T.pack . uriFragment <$> dynLoc
+  dynLoc <- browserHistoryWith getLocationUri
+  let dynFrag = decodeFrag . T.pack . uriFragment <$> dynLoc
 
-    -- ticker1 <- tickLossy 1 =<< liftIO getCurrentTime
+  -- ticker1 <- tickLossy 1 =<< liftIO getCurrentTime
 --    ticker1cnt <- count ticker1
 
-    -- top-level PB event
-    evPB0 <- getPostBuild
-
-    t0 <- liftIO getCurrentTime
-
-    dynClock <- clockLossy 1 t0
-    let dynUTCTime = _tickInfo_lastUTC <$> dynClock
+  -- top-level PB event
+  evPB0 <- getPostBuild
+
+  t0 <- liftIO getCurrentTime
+
+  dynClock <- clockLossy 1 t0
+  let dynUTCTime = _tickInfo_lastUTC <$> dynClock
+
+  ticker2 <- tickLossy 2 t0
+  let ticker4 = ffilter ((==0) . flip rem 2 . _tickInfo_n) ticker2
+      ticker8 = ffilter ((==0) . flip rem 4 . _tickInfo_n) ticker2
+
+  -- we can use this
+  evIdxStLast <- getIdxStates (leftmost [ticker8 $> (), evPB0 $> ()])
+  dynIdxStLast <- holdUniqDyn =<< holdDyn (PkgIdxTs 0) evIdxStLast
+
+  evPackages0 <- getPackages (updated dynIdxStLast $> ())
+  dynPackages0 <- holdDyn mempty evPackages0
+
+  -- pseudo navbar
+  el "nav" $ do
+    text "[ "
+    elAttr "a" ("href" =: "#/") $ text "HOME"
+    text " | "
+    elAttr "a" ("href" =: "#/queue") $ text "Build Queue"
+    text " | "
+    elAttr "a" ("href" =: "#/packages") $ text "Packages"
+    text " ]"
+    text "    (current index-state: "
+    dynText (pkgIdxTsToText <$> dynIdxStLast)
+    text " ; package count: "
+    display (V.length <$> dynPackages0)
+    text ")"
+
+  -- search box
+  searchInputE <- elAttr "div" ("class" =: "item search right clearfix") $ do
+            divClass "text" $ text "Package Search"
+            sVal0 <- inputElement $ def & inputElementConfig_elementConfig . elementConfig_initialAttributes .~ fold
+              ["class" =: "input-search","placeholder" =: "search..."]
+            debounce 0.1 $ _inputElement_input sVal0
+          
+  let
+    dynPackagesJss = V.toList . fmap (JSS.textToJSString . pkgNToText) <$> dynPackages0
+    calcMatches pkgs sJss =
+      if JSS.length sJss < 3
+      then []
+      else matches' pkgs sJss
+     where 
+      matches' p sJ
+        | JSS.isInfixOf "^" sJ = filter (JSS.isPrefixOf (JSS.dropWhile (=='^') sJ)) p
+        | JSS.isInfixOf "$" sJ = filter (JSS.isSuffixOf (JSS.dropWhileEnd (=='$') sJ)) p
+        | otherwise            = filter (JSS.isInfixOf sJ) p
+    exactMatches pkgs' sJss = List.partition (==(JSS.dropWhileEnd (=='$') sJss)) pkgs'
+    matchesE = calcMatches <$> current dynPackagesJss <@> (JSS.textToJSString <$> searchInputE)
+  dynMatches <- holdDyn [] matchesE
+  dynSearch  <- holdDyn "" (JSS.textToJSString <$> searchInputE)
```

Comment:
I suggest defining a 2nd `dynSearchFiltered` with any `$` or `^` removed; and use that one below in the code for the `<strong>`-splitting